File: 2_ReactWeb/src/features/tool-market/model/toolMarket.ts
export type ToolMarketInstallationStatus =
  | "not-installed"
  | "installed"
  | "update-available"
  | "call-name-conflict";

export type ToolMarketTool = {
  author: string;
  callName: string;
  compatibility: { minTianceVersion: string; platforms: string[] };
  displayName: string;
  id: string;
  installationStatus: ToolMarketInstallationStatus;
  license: string;
  localCallName: string | null;
  localProjectId: string | null;
  localVersion: string | null;
  packageUrl: string;
  runtime: string;
  sha256: string;
  size: number;
  suggestedCallName: string | null;
  summary: string;
  version: string;
};

export type ToolMarketFilters = {
  authors: string[];
  platforms: string[];
  runtimes: string[];
  statuses: ToolMarketInstallationStatus[];
};

export type ToolMarketSettings = {
  filters: ToolMarketFilters;
  schemaVersion: 1;
  source: string;
};

export type ToolMarketIndex = {
  cached: boolean;
  kind: "tiance-tool-market";
  name: string;
  schemaVersion: 1;
  source: string;
  tools: readonly ToolMarketTool[];
  updatedAt: string;
};

export function filterToolMarketItems(
  tools: readonly ToolMarketTool[],
  filters: ToolMarketFilters,
  query: string,
) {
  const normalizedQuery = query.trim().toLocaleLowerCase();
  return tools.filter((tool) => {
    if (filters.authors.length && !filters.authors.includes(tool.author)) return false;
    if (filters.runtimes.length && !filters.runtimes.includes(tool.runtime)) return false;
    if (
      filters.platforms.length
      && !filters.platforms.some((platform) => tool.compatibility.platforms.includes(platform))
    ) return false;
    if (filters.statuses.length && !filters.statuses.includes(tool.installationStatus)) return false;
    if (!normalizedQuery) return true;
    return [tool.displayName, tool.callName, tool.id, tool.author, tool.summary, tool.runtime]
      .some((value) => value.toLocaleLowerCase().includes(normalizedQuery));
  });
}

export function listToolMarketValues(
  tools: readonly ToolMarketTool[],
  key: "author" | "runtime",
) {
  return [...new Set(tools.map((tool) => tool[key]))]
    .sort((left, right) => left.localeCompare(right));
}

export function listToolMarketPlatforms(tools: readonly ToolMarketTool[]) {
  return [...new Set(tools.flatMap((tool) => tool.compatibility.platforms))]
    .sort((left, right) => left.localeCompare(right));
}

File: 1_PythonServer/app/schemas/tools/tool_market.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import BaseModel, ConfigDict, Field


ToolMarketInstallationStatus = Literal[
    "not-installed",
    "installed",
    "update-available",
    "call-name-conflict",
]


class ToolMarketContract(BaseModel):
    model_config = ConfigDict(populate_by_name=True, extra="forbid")


class ToolMarketFilterSettings(ToolMarketContract):
    authors: list[str] = Field(default_factory=list)
    runtimes: list[str] = Field(default_factory=list)
    platforms: list[str] = Field(default_factory=list)
    statuses: list[ToolMarketInstallationStatus] = Field(default_factory=list)


class ToolMarketSettingsResponse(ToolMarketContract):
    schema_version: Literal[1] = Field(default=1, alias="schemaVersion")
    source: str
    filters: ToolMarketFilterSettings = Field(default_factory=ToolMarketFilterSettings)


class ToolMarketSettingsUpdateRequest(ToolMarketContract):
    filters: ToolMarketFilterSettings


class ToolMarketConnectRequest(ToolMarketContract):
    source: str


class ToolMarketCompatibility(ToolMarketContract):
    min_tiance_version: str = Field(alias="minTianceVersion")
    platforms: list[str]


class ToolMarketEntry(ToolMarketContract):
    id: str
    version: str
    author: str
    license: str
    call_name: str = Field(alias="callName")
    display_name: str = Field(alias="displayName")
    summary: str
    runtime: str
    package_url: str = Field(alias="packageUrl")
    sha256: str
    size: int
    compatibility: ToolMarketCompatibility


class ToolMarketRemoteIndex(ToolMarketContract):
    schema_version: Literal[1] = Field(alias="schemaVersion")
    kind: Literal["tiance-tool-market"]
    name: str
    updated_at: datetime = Field(alias="updatedAt")
    tools: list[ToolMarketEntry]


class ToolMarketToolResponse(ToolMarketEntry):
    installation_status: ToolMarketInstallationStatus = Field(alias="installationStatus")
    local_project_id: str | None = Field(default=None, alias="localProjectId")
    local_version: str | None = Field(default=None, alias="localVersion")
    local_call_name: str | None = Field(default=None, alias="localCallName")
    suggested_call_name: str | None = Field(default=None, alias="suggestedCallName")


class ToolMarketIndexResponse(ToolMarketContract):
    schema_version: Literal[1] = Field(alias="schemaVersion")
    kind: Literal["tiance-tool-market"]
    name: str
    updated_at: datetime = Field(alias="updatedAt")
    source: str
    cached: bool
    tools: list[ToolMarketToolResponse]


class ToolMarketInstallRequest(ToolMarketContract):
    category_id: str | None = Field(default=None, alias="categoryId")
    call_name: str | None = Field(
        default=None,
        alias="callName",
        pattern=r"^[a-z][a-z0-9_]*$",
    )


class ToolMarketInstallResponse(ToolMarketContract):
    tool_id: str = Field(alias="toolId")
    project_id: str = Field(alias="projectId")
    category_id: str = Field(alias="categoryId")
    call_name: str = Field(alias="callName")
    version: str
    updated: bool
    has_dependencies: bool = Field(alias="hasDependencies")


class ToolPackageAuthor(ToolMarketContract):
    name: str


class ToolPackageManifest(ToolMarketContract):
    schema_version: Literal[1] = Field(alias="schemaVersion")
    kind: Literal["tiance-tool-package"]
    id: str
    version: str
    author: ToolPackageAuthor
    license: str
    compatibility: ToolMarketCompatibility
